[2단계 - 로또 리팩토링] 포포(이민우) 미션 제출합니다.

Minuring opened the pull request:
아서의 피드백을 듣고 아래와 같이 적용해 보았습니다!

# 첫 번째 피드백 이후 변경사항
### 1. 핵심 비즈니스 로직을 가지는 객체를 domain 패키지로 분류
- 도메인 객체들의 패키지명을 model에서 domain으로 변경하였습니다.

### 2. 로직을 단순히 위임만 하는 서비스 제거
- 컨트롤러가 직접 도메인의 로직을 호출하도록 변경하였습니다.
단순히 도메인 로직의 호출을 위임만 하던 `LottoEvaluationService`는 불필요하다고 판단하였습니다.

### 3. 단순히 데이터 복사를 하던 DTO들을 제거
- 단순히 도메인 객체의 데이터에 대한 복사본으로 사용되던 DTO들을 제거하되, 원본 객체를 변경할 수 없도록 도메인 객체를 불변으로 설계하였습니다.

### 4. 도메인 객체를 생성하는 별도의 Factory 제거
- 특별한 복잡한 생성 로직이 없기 때문에 Factory는 오버엔지니어링이라고 생각하여 제거했습니다.

### 5. 당첨 로또에 대해 싱글톤 패턴 구현을 철회
- 가장 큰 이유로 테스트 하기 어렵다는 점, 그리고 이미 `Lotto`와 `WinningLotto`가 별도로 존재하기 때문에 "당첨 로또"라는 의미를 살릴 수 있다고 생각하였습니다.
당첨 로또를 사용하는 쪽에서 한 인스턴스만 관리한다면 "유일한 당첨 로또"라는 의미 또한 전달할 수 있다고 판단하였습니다.

<hr>

## 2월 19일 리뷰 요청 변경 사항
아서가 남겨준 피드백 덕분에 한 층 성장하는 것 같습니다!
소중한 피드백 정말 정말 감사합니다 ㅎㅎ 많은 걸 배웠습니다!!

1. 로또번호 생성 전략 인터페이스 이름을 명확하게 변경했습니다.
- NumbersStretagy -> LottoNumbersGenerator

2. 랜덤한 로또 번호 생성을 유틸이 아닌 전략 인터페이스 구현체로 변경했습니다.
- (제거) util.RandomGenerator 
- (추가) service.RandomLottoNumbersGenerator

3. 로또 번호를 객체로 승격시켰습니다.
- -> int형을 사용했을 때 불가능했던 제한적인 범위와 같은 로또 번호에 대한 제약을 추가할 수 있었습니다. 또한 해당 제약에 대한 중복된 검증을 제거할 수 있었습니다.

4. 로또 당첨 통계와 관련한 클래스들을 구체적인 이름으로 변경했습니다.
- Statistics -> LottoStatisticsDto
- StatisticsService -> LottoStatisticsService

5. 가독성을 해치는 무분별한 static import를 제거했습니다.
6. 불필요한 클래스들을 제거했습니다.
- ViewFacade, Container

Minuring committed:
chore: model 패키지명 변경

요구사항에 따라 도메인 패키지명을 "model"에서 "domain"으로 변경합니다.

Minuring committed:
refactor(service): 로또 발행 시 번호 생성 전략을 주입하도록 변경

정상적으로 로또 번호가 발행 되었는 지에 대한 테스트를 가능하도록 하기 위해 전략 패턴을 사용하도록 변경합니다.

Minuring committed:
refactor(dto): 도메인을 불변 객체로 하여 DTO 제거

도메인 객체를 불변하게 설계하고, 단순히 도메인의 데이터를 복사하던 DTO를 제거하고 도메인을 바로 사용하도록 변경합니다.

Minuring committed:
refactor(dto): 당첨 통계를 DTO로 분류

당첨 통계를 나타내는 데이터로 쓰이던 Statistics를 도메인이 아닌 DTO로 분류합니다.

Hyunta commented on src/main/java/service/StatisticsService.java:
```diff
@@ -2,12 +2,12 @@
-import model.PrizeTier;
-import model.Statistics;
 
 public class StatisticsService {
```
StatisticsService를 외부에서 LottoController에 주입하신 이유가 있을까요?

Minuring replied:
이유없이 관성적으로 의존성을 주입한다는 접근을 했던 것 같습니다..!🙄

Hyunta replied:
아하 그럴수 있죠!
그래도 지금은 학습하는 단계이니 왜? 라는 질문을 던져보면 성장하시는데 더 도움이 되실 것 같습니다ㅎㅎ

Hyunta commented on src/main/java/view/ViewFacade.java:
```diff
@@ -1,7 +1,7 @@
+import domain.Lottos;
+import dto.Statistics;
 
 public class ViewFacade {
```
ViewFacade는 어떤 필요로 만들게 되었을까요?

Minuring replied:
컨트롤러에서 InputView, OutputView를 구분하지 않고 "뷰"로만 바라보고싶은 생각을 했었습니다.
불필요한 구현으로 오히려 의도를 알 수 없는 코드가 되어버린 것 같습니다!

Hyunta replied:
의도가 있다면 어떤 접근이든 저는 가치가 있다고 생각합니다. 👍🏻
질문을 드렸던 의도는 controller 입장에서는 viewFacade만 바라보면 되니까 단순할 수 있습니다.
하지만 코드를 유지보수하는 관리자 입장에서는 파사드로 묶여있는 클래스가 항상 장점만 가지지는 않습니다.
하는 일이 너무 복잡해서 만들었다면 의미를 가졌을 것 같은데 현재는 개인적으로 단순한 기능이라고 생각해서 클래스를 나누지 않아도 괜찮을 거라는 생각이 들어서 질문 드렸습니다.

Hyunta commented on src/main/java/controller/LottoController.java:
```diff
@@ -33,8 +34,9 @@ public void run() {
-        Lottos lottos = lottoGenerateService.generateLottos(purchaseAmount);
-        viewFacade.printLottos(lottos.toDto());
+        Lottos lottos = lottoGenerateService.generateLottos(purchaseAmount,
+            () -> RandomGenerator.generateNumbers(1, 45, 6));
```
RandomGenerator를 util 패키지 하위로 놓으셨는데 유틸성이라고 생각하신 이유가 있을까요?
정해진 카운트만큼 랜덤 값을 가져오는 것은 로또의 비즈니스 로직으로 볼 수 있지 않나요?

Minuring replied:
그러네요! 1~45 사이 범위, 음수 불가와 같은 제약이 존재하는 이상 유틸이 아니라고 생각됩니다

Hyunta commented on src/main/java/dto/Statistics.java:
```diff
@@ -1,14 +1,14 @@
-import model.PrizeTier;
 
-public class StatisticsDto {
+public class Statistics {
```
이름을 조금 구체적으로 가져가면 좋겠네요, 어떤 Statistics 인가요?
그리고 저는 Dto 임을 명시해주는 편이 좋아보여요, 동료 개발자들은 이게 Dto인지 알려면 클래스이름 뿐만아니라 패키지명도 살펴봐야하니까요.

Minuring replied:
제가 봐도 그렇네요!
이름을 너무 모호하게 지은 것 같습니다

Hyunta commented on src/main/java/controller/LottoController.java:
```diff
@@ -33,8 +34,9 @@ public void run() {
 
     private PurchaseHistory processLottoPurchase() {
         int purchaseAmount = convertToInteger(viewFacade.getPurchaseInput());
```
convertToInteger는 어느 클래스에 있는 메서드일까요?

static 메서드인지 내부 메서드인지 불분명하네요.
static 메서드의 import static은 지양해주세요.

아래 스택플로우 글을 한번 읽어보시면 도움될 것 같네요ㅎㅎ
https://stackoverflow.com/questions/420791/what-is-a-good-use-case-for-static-import-of-methods

Minuring replied:
단순히 코드 양을 줄이려고 static import를 했습니다..ㅎㅎ
말씀 듣고 보니 코드 작성자 자신만 알아볼 수 있겠네요!
글 잘 읽었습니다 😊

Hyunta commented on src/main/java/service/NumbersStrategy.java:
```diff
@@ -0,0 +1,7 @@
+import java.util.List;
+import java.util.function.Supplier;
+
+public interface NumbersStrategy extends Supplier<List<Integer>> {
+}
```
해당 클래스의 이름을 조금더 구체적으로 작성해주세요.

Hyunta reviewed (changes requested):
안녕하세요 포포,
2단계 리팩터링 미션도 잘 진행해주셨네요.
몇가지 커멘트 남겼으니 확인 부탁드립니다.

Minuring committed:
refactor: 로또번호 생성 전략 인터페이스 이름 수정

단순 숫자들이 아닌 로또 번호를 생성한다는 의미를 전달하기 위해 이름을 변경합니다.

Minuring committed:
refactor: 랜덤 로또번호 생성을 유틸 대신 전략 구현체로 변경

랜덤 로또 번호 생성은 아래와 같은 비즈니스 규칙이 있기 때문에 유틸이 아닌 비즈니스 로직으로 분류합니다.
- 1~45 사이
- 6개
- 중복 없음

Minuring committed:
refactor: "로또 번호" 객체 생성

로또와 당첨로또에서 숫자 범위 검증이 중복되던 점,
입력을 정수로 변환하는 유틸에서 숫자 범위에 대한 검증을 수행한 점을 수정하기 위해
정수를 포장한 객체로 로또 번호를 표현합니다.

Minuring committed:
refactor: 로또 당첨 통계 서비스와 DTO에 대해 구체적인 이름으로 변경

Statistics보다 LottoStatisticsDto가,
StatisticsService보다 LottoStatisticsService가 더 명확하게 전달되는 이름이라고 판단하여 변경합니다.

Minuring committed:
refactor: 가독성을 해치는 static import 제거

Minuring committed:
refactor(view): 불필요한 ViewFacade 제거

단순히 뷰 호출을 위임하기 위해서라면 필요하지 않다고 판단해 제거합니다.

Minuring committed:
refactor: 의존성 관리 목적의 컨테이너 제거

관리해야할 만큼 객체들 간의 의존성이 복잡하지 않기 때문에 제거합니다.

Minuring committed:
refactor: 의존성 관리 목적의 컨테이너 제거

관리해야할 만큼 객체들 간의 의존성이 복잡하지 않기 때문에 제거합니다.

Hyunta reviewed (approved):
안녕하세요 포포,
2단계 리팩터링 미션도 잘 진행해주셨습니다.
남겨주신 커멘트에 답변 남겼는데 혹시 더 궁금하신 점이 있으면 편하게 DM 주세요.
남은 미션도 응원하겠습니다!
